// @genkit-ai/vertexai/lib/modelgarden/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var modelgarden_exports = {};
__export(modelgarden_exports, {
  claude35Sonnet: () => import_legacy.claude35Sonnet,
  claude35SonnetV2: () => import_legacy.claude35SonnetV2,
  claude3Haiku: () => import_legacy.claude3Haiku,
  claude3Opus: () => import_legacy.claude3Opus,
  claude3Sonnet: () => import_legacy.claude3Sonnet,
  claudeOpus4: () => import_legacy.claudeOpus4,
  claudeOpus41: () => import_legacy.claudeOpus41,
  claudeSonnet4: () => import_legacy.claudeSonnet4,
  codestral: () => import_legacy.codestral,
  llama3: () => import_legacy.llama3,
  llama31: () => import_legacy.llama31,
  llama32: () => import_legacy.llama32,
  mistralLarge: () => import_legacy.mistralLarge,
  mistralNemo: () => import_legacy.mistralNemo,
  vertexAIModelGarden: () => import_legacy.vertexAIModelGarden,
  vertexModelGarden: () => import_v2.vertexModelGarden
});
module.exports = __toCommonJS(modelgarden_exports);
var import_legacy = require("./legacy/index.js");
var import_v2 = require("./v2/index.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  claude35Sonnet,
  claude35SonnetV2,
  claude3Haiku,
  claude3Opus,
  claude3Sonnet,
  claudeOpus4,
  claudeOpus41,
  claudeSonnet4,
  codestral,
  llama3,
  llama31,
  llama32,
  mistralLarge,
  mistralNemo,
  vertexAIModelGarden,
  vertexModelGarden
});
