// @genkit-ai/vertexai/lib/rerankers/v2/index.mjs
import {
  genkitPluginV2
} from "genkit/plugin";
import { getDerivedOptions } from "../../common/utils.js";
import * as reranker from "./reranker.js";
async function initializer(pluginOptions) {
  const clientOptions = await getDerivedOptions(
    "vertex-rerankers",
    pluginOptions
  );
  return await reranker.listKnownRerankers(clientOptions);
}
async function resolver(actionType, actionName, pluginOptions) {
  const clientOptions = await getDerivedOptions(
    "vertex-rerankers",
    pluginOptions
  );
  if (actionType == "reranker" && reranker.isRerankerModelName(actionName)) {
    return reranker.defineReranker(actionName, clientOptions);
  }
  return void 0;
}
function vertexRerankersPlugin(options) {
  return genkitPluginV2({
    name: "vertex-rerankers",
    init: async () => await initializer(options),
    resolve: async (actionType, actionName) => await resolver(actionType, actionName, options)
  });
}
const vertexRerankers = vertexRerankersPlugin;
vertexRerankers.reranker = (name, config) => {
  return reranker.reranker(name, config);
};
export {
  vertexRerankers,
  vertexRerankersPlugin
};
