// @genkit-ai/vertexai/lib/rerankers/v2/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var v2_exports = {};
__export(v2_exports, {
  vertexRerankers: () => vertexRerankers,
  vertexRerankersPlugin: () => vertexRerankersPlugin
});
module.exports = __toCommonJS(v2_exports);
var import_plugin = require("genkit/plugin");
var import_utils = require("../../common/utils.js");
var reranker = __toESM(require("./reranker.js"));
async function initializer(pluginOptions) {
  const clientOptions = await (0, import_utils.getDerivedOptions)(
    "vertex-rerankers",
    pluginOptions
  );
  return await reranker.listKnownRerankers(clientOptions);
}
async function resolver(actionType, actionName, pluginOptions) {
  const clientOptions = await (0, import_utils.getDerivedOptions)(
    "vertex-rerankers",
    pluginOptions
  );
  if (actionType == "reranker" && reranker.isRerankerModelName(actionName)) {
    return reranker.defineReranker(actionName, clientOptions);
  }
  return void 0;
}
function vertexRerankersPlugin(options) {
  return (0, import_plugin.genkitPluginV2)({
    name: "vertex-rerankers",
    init: async () => await initializer(options),
    resolve: async (actionType, actionName) => await resolver(actionType, actionName, options)
  });
}
const vertexRerankers = vertexRerankersPlugin;
vertexRerankers.reranker = (name, config) => {
  return reranker.reranker(name, config);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  vertexRerankers,
  vertexRerankersPlugin
});
